runtime/src/bank/builtins/core_bpf_migration/source_bpf_upgradeable.rs

```diff
@@ -0,0 +1,365 @@
+use {
+    super::error::CoreBpfMigrationError,
+    crate::bank::Bank,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        bpf_loader_upgradeable::{
+            get_program_data_address, UpgradeableLoaderState, ID as BPF_LOADER_UPGRADEABLE_ID,
+        },
+        pubkey::Pubkey,
+    },
+};
+
+/// Used to validate a source BPF Upgradeable program's account and data
+/// account before migrating a built-in program to Core BPF.
+#[derive(Debug)]
+pub(crate) struct SourceProgramBpfUpgradeable {
+    pub program_address: Pubkey,
+    pub program_account: AccountSharedData,
+    pub program_data_address: Pubkey,
+    pub program_data_account: AccountSharedData,
+    pub total_data_size: usize,
+}
+
+impl SourceProgramBpfUpgradeable {
+    fn check_program_account(&self) -> Result<(), CoreBpfMigrationError> {
+        // The program account should be owned by the upgradeable loader.
+        if self.program_account.owner() != &BPF_LOADER_UPGRADEABLE_ID {
+            return Err(CoreBpfMigrationError::IncorrectOwner(self.program_address));
+        }
+
+        // The program account should have a pointer to its data account.
+        if let UpgradeableLoaderState::Program {
+            programdata_address,
+        } = &self
+            .program_account
+            .deserialize_data()
+            .map_err::<CoreBpfMigrationError, _>(|_| {
+                CoreBpfMigrationError::InvalidProgramAccount(self.program_address)
```

Comment:
Hmm, I don't love that this eats the deserialize error. That's a very different situation than the `programdata_address` being the wrong value. Can you add a deserialization error variant that preserves the bincode error text?

Reply:
Sure. I lumped "failed to deserialize" and "state is valid but it's got the wrong value(s)" into the same "invalid" error. I figured in either case the state is bad, but I see your point; it's useful to distinguish between the two. I'll add a bincode error for specifically deserialization failures.

Note this will eliminate the ability to use `PartialEq` on the error, which may or not be a big deal.

Reply:
> Note this will eliminate the ability to use `PartialEq` on the error, which may or not be a big deal.

Seems fine to me.